nemo/collections/nlp/models/nlp_model.py

```diff
@@ -312,53 +252,225 @@ def setup(self, stage: str) -> None:
-                        # update path based on model parallel rank
-                        filepath = self._trainer.resume_from_checkpoint
-                        dirname = os.path.dirname(os.path.dirname(filepath))
-                        basename = os.path.basename(filepath)
-                        filepath = f'{dirname}/mp_rank_{app_state.model_parallel_rank:02d}/{basename}'
-                        self._trainer.resume_from_checkpoint = filepath
-                        logging.info(f'Resuming training from checkpoint {self._trainer.resume_from_checkpoint}')
-                        # need to set checkpoint version for megatron-lm
-                        checkpoint_version = torch.load(self._trainer.resume_from_checkpoint).get(
-                            'checkpoint_version', None
-                        )
-                        if checkpoint_version is not None:
-                            set_checkpoint_version(checkpoint_version)
-                        else:
-                            logging.warning(
-                                'Megatron-lm checkpoint version not found. Setting checkpoint_version to 0.'
-                            )
-                            set_checkpoint_version(0)
-                    else:
-                        logging.info(
-                            f"Restoring from pretrained model parallel checkpoint: {self.bert_model._restore_path}"
-                        )
-                        self.bert_model.restore_weights(self.bert_model._restore_path)
-
-                    logging.info("Replacing sampler with model parallel sampler")
-                    mp_sampler = torch.utils.data.distributed.DistributedSampler(
-                        self._train_dl.dataset,
-                        num_replicas=app_state.data_parallel_size,
-                        rank=app_state.data_parallel_rank,
-                    )
-                    mp_dl = self._trainer.replace_sampler(self._train_dl, mp_sampler)
-                    self._train_dl = mp_dl
-                else:
-                    raise NotImplementedError(
-                        f'The BERT encoder: {self.bert_model} does not support model parallelism yet.'
-                    )
-
     def on_save_checkpoint(self, checkpoint: Dict[str, Any]) -> None:
         if hasattr(self, "bert_model") and isinstance(self.bert_model, MegatronBertEncoder):
             checkpoint['checkpoint_version'] = get_checkpoint_version()
         return None
 
+    def on_load_checkpoint(self, checkpoint: Dict[str, Any]) -> None:
+        if hasattr(self, "bert_model") and isinstance(self.bert_model, MegatronBertEncoder):
+            set_checkpoint_version(checkpoint['checkpoint_version'])
+            logging.info(f"Setting Megatron checkpoint version: {checkpoint['checkpoint_version']}")
+        return None
+
+    # no rank check as model parallel models need to be saved on data parallel rank 0
```

Comment:
Its true that model parallel models dont need to be saved on rank 0, *non* model parallel models would suffer issues we had before - all nodes try to save to same location and it corrupts the file. Its fine to remove the decorator but have an equivalent such that when 1) its a non-model parallel model 2) its any rank != 0 ;; immediately return and dont try to save anything.

Reply:
Or is this implicitly handled by `super().save_to()`? In that case, disregard above comment.

Reply:
That's correct. `super().save_to()` has the decorator for the non-model parallel case.

Reply:
Ok perfect then